### ISSUE TITLE
Mega Issue: Extend Cloud Providers

### ISSUE THREAD

**Issue by jonathan-innis** [aws/karpenter#2532]
### Tell us about your request

There's a growing number of requests to extend the cloud providers that Karpenter supports from AWS to other cloud providers (Azure, GCP, Orcale, etc.). This issue is intended to cover the discussion of extending the number of cloud providers broadly.

### Tell us about the problem you're trying to solve. What are you trying to do, and why is it hard?

N/A

### Are you currently working around this issue?

N/A

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by ellistarn** [comment]
I'd like to explore two models for this:
1: Karpenter as a Library (as currently implemented here: https://github.com/bwagner5/karpenter-k3d)
2: Karpenter as an API: https://github.com/kubernetes-csi

A few table stakes in my mind:
1. Vendors should be able to define their own API surface (e.g. AWSNodeTemplate) that's separately versioned
2. Vendor code should live in a separate repository from Karpenter core (k8s is doing this across the board w/ in-tree / out-of-tree)

**Comment by debugger24** [comment]
Happy to work on this. Let me know how we can colabrate.

**Comment by jonathan-innis** [comment]
@debugger24 Are you interested in helping/building a particular provider? GCP? Oracle? Digital Ocean?

**Comment by tonymarkel** [comment]
I'd like to help with the Oracle portions of this in whatever capacity I can. If I read @ellistarn correctly, I'm okay to create a separate repo like @bwagner5 did. I haven't been able to find an equivalent open-source project on the 'net, so hopefully I'm not re-inventing a wheel. If I don't see anything in this thread in a day or so, I'll follow Brandon's example and put what little I have in a repo and share it here.

**Comment by jonathan-innis** [comment]
Nope! @tonymarkel. Feel free to take a stab at the Oracle cloudprovider and share the repo in this issue!

**Comment by ellistarn** [comment]
The best example you'll find is to look at aws/karpenter as the aws implementation. You'll need your provider to rely on aws/karpenter-core.

**Comment by sftim** [comment]
Should we label this with _neutrality_?

**Comment by Anhui-tqhuang** [comment]
seems that if we want to make a cloud provider, we cannot reuse the existing repo as it is tied to AWS implementation, we need to use karpenter-core to create our own karpenter ....
